File: patterns/stack/stack_of_intermediate_results/is_valid.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <unordered_map>
#include "./utils.h"

using namespace std;

bool is_valid(const vector<char> &symbols)
{
    // Stage 1: Initialization
    stack<char> _stack;
    unordered_map<char, char> patterns = {{'(', ')'}, {'[', ']'}, {'{', '}'}, {'<', '>'}};

    // Stage 2: Main loop
    for (char symbol : symbols)
    {
        if (patterns.contains(symbol))
            _stack.push(symbol); // Is opening symbol since is in the map, so add to stack and continue
        else if (_stack.empty())
            return false; // Is a closing symbol since is no in the map, so is not valid
        else
        {
            char head = _stack.top(); // Obtain the head element
            _stack.pop();             // Now remove the head element since we have its value

            if (symbol != patterns[head])
                return false;
        }
    }
    return _stack.empty();
}

int main()
{
    vector<char> symbols = {'(', ')', '<', '[', '{', '}', ']', '>'};
    vector<char> bad_symbols = {'(', '(', ')'};

    pprint(symbols);
    cout << "Is valid: " << (is_valid(symbols) ? "True" : "False") << endl;

    pprint(bad_symbols);
    cout << "Is valid: " << (is_valid(bad_symbols) ? "True" : "False") << endl;

    return 0;
}

File: patterns/stack/stack_of_intermediate_results/is_valid.py
"""
We are given with a list of symbols, we need to check if every opening symbol
has its closing symbol.

e.g.
symbols = ["(","(", "<", "[","{","}","]",">"]
output = True

symbols = ["(","(",")"]
output = False

This is a stack problem, so we need to use a stack to keep track of the opening symbols.
"""


def is_good(symbols: list[str]) -> bool:
    stack: list[str] = []

    patterns: dict[str, str] = {"(": ")", "[": "]", "{": "}", "<": ">"}

    for symbol in symbols:
        if symbol in patterns:
            # CHAR is an opening symbol, so we add it to the stack
            stack.append(symbol)
        elif len(stack) == 0:
            # CHAR is a closing symbol, but without openning one, so finish that
            return False
        else:
            # CHAR is closing one, but we have elements in the stack, so we check
            top = stack.pop()
            if symbol != patterns[top]:
                return False

    return len(stack) == 0


if __name__ == "__main__":
    symbols: list[str] = ["(", ")", "<", "[", "{", "}", "]", ">"]
    bad_symbols: list[str] = ["(", "(", ")"]

    print(f"Symbols: {symbols} -> {is_good(symbols)}")
    print(f"Bad symbols: {bad_symbols} -> {is_good(bad_symbols)}")
